### ISSUE TITLE
Repo size is many times larger than the sum of all deduped/compressed file sizes... Is this normal?

### ISSUE THREAD

**Issue by tnyeanderson** [borgbackup/borg#7118]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Question, possible bug report

#### Your borg version (borg -V).

Client running the backups is using 1.2.1 using the `b3vis/borgmatic` docker image.

Using borg v1.2.2 locally to run `borg info`.

Server is BorgBase

#### Operating system (distribution) and version.

See above

#### Hardware / network configuration, and filesystems used.

Data is bind mounted into the container.

#### How much data is handled by borg?

That's the real question :) between 4 and 7TB. There is a little over 4TB of data that I know is currently backed up (not sure of the deduped size on it though).

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

See problem description.

## Describe the problem you're observing.

If I use `borg info`, it says:

```
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
All archives:              138.34 TB            129.69 TB              7.01 TB

                       Unique chunks         Total chunks
Chunk index:                 3481132             65133513
```

That deduplicated size seems high (honestly all the sizes seem high) so I ran my own test. I wrote a script which iterates through all the archives in a repo and lists all the files in each one, effectively listing every reference to every file in every archive in the repo. I also print the file's `dcsize` (deduped/compressed size) before each path.

`list-all-files-in-repo.sh`:
```
#!/bin/bash

repo=/path/to/myrepo

BORG_PASSCOMMAND="cat $HOME/.borgpass"
export BORG_PASSCOMMAND

for archive in $(borg list "$repo" --format='{name} '); do
        borg --consider-part-files list "$repo::$archive" --format='{dcsize} {path}{NEWLINE}'
done
```

The output of this script looks like:
```
0 mnt/source/path/to/file
2333421 mnt/source/path/to/file2
...
```

I then took this output and summed all of these "first column" numbers:
```
$ ./list-all-files-in-repo.sh | awk '{print $1}' | paste -sd'+' - | bc
1197793067918
```

If this answer is in bytes, then it equals `1115.53 GB`, far less than the `7.01 TB` found with `borg info`. I think all my work is correct, in that the final number should include the final size of all the deduped/compressed data within the repo. What else could be taking up this space?

I've run `borg compact --cleanup-commits` to try to clean up any unused data, but the answer is the same.

I am using borgmatic to perform the backups. It automatically prunes the archives. The repo currently has 37 archives.

Using borg v1.2.2 to run `borg info`.

**TLDR:**

- Is `dcsize` in bytes? Or chunks? Or something else?
- Should the sum of all `dcsize`s of all files in all archives in a repo be equal to the space shown in the `deduplicated` column for `All archives` in `borg info`?
- How can I find what is taking up all the space in my repo?

Happy to provide any other info if needed. Thanks in advance!

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

Size remains after compact. See above.

#### Include any warning/errors/backtraces from the system logs

None

**Comment by infectormp** [comment]
Maybe you are using `append-only` mode for your repo?

**Comment by ThomasWaldmann** [comment]
Please also add the `du -sh repo_directory`.

**Comment by tnyeanderson** [comment]
@infectormp You are correct (in part). The borgmatic client that performs the backups is using a key that is set as append only. However my local key (that I am using for `borg info` and `borg compact`) is set to "full access". Could that be contributing to the issue I'm experiencing?

@ThomasWaldmann So I did `borg mount $repo borg-repo` and then ran `du -sh borg-repo`. This gives me `113 TB`... not sure if this is what you're looking for. I don't have shell access to the server (borgbase) so I can't run it directly... and I suppose there is magic happening in the mount that doesn't represent deduped data properly.

**Comment by ThomasWaldmann** [comment]
I meant the disk usage of the repo dir, not of the borg mount.

**Comment by tnyeanderson** [comment]
That's what I figured. Do you happen to know if there is a way for me to get that info through borgbase? Would a standard `sshfs` mount of the repo accomplish this? I've had trouble doing a plain sshfs mount in the past but could try again if it will help :)

**Comment by ThomasWaldmann** [comment]
Yes, guess sshfs would give the wanted results. I mean the repository disk usage (not the source folders).

Basically the **real** fs disk usage, not borg's internal bookkeeping.

**Comment by tnyeanderson** [comment]
BorgBase was kind enough to run the needed command for me :)

```
/srv/repos/myrepo # du -sh repo
6.4T    repo
```

**Comment by ThomasWaldmann** [comment]
OK, so let's analyse this:
- dcsize is in bytes and it means the size of the chunks **only** referenced from that archive/file.
- while dcsize is interesting (as in "what would be freed if we deleted this"), there are of course also chunks referenced by **multiple** archives/files (some even by many or all archives) - these will nowhere show up in your dcsize listing.
- 7.0TB (borg value, uses 10^12) == 6.4TiB (Linux du value, uses 2^40)

Finding out what takes space:
- if one keeps the borg create --list --stats logs, one could see jumps in deduped size for the **first** archive created with lots of new stuff
- but as soon as you make another backup with the same stuff, it will already dedupe against the previous backup.
- borg info of an archive always shows the dedup size against the rest of the repo (all other archives)
- borg diff maybe useful

**Comment by tnyeanderson** [comment]
Ah I think I see the issue in my analysis then!

>while dcsize is interesting (as in "what would be freed if we deleted this"), there are of course also chunks referenced by multiple archives/files (some even by many or all archives) - these will nowhere show up in your dcsize listing.

So essentially, all that additional space is being taken up by *shared* chunks, which aren't represented at all in `dcsize`. Is the proper way of calculating the total size of shared chunks essentially `borg_dedup_size - total_dc_size`, e.g. `6.4TB - 1.12TB`?

If that's correct, then I think this can be counted as closed as my question has been answered!

I can try using `borg diff` to find out where the size increases come from. Just to confirm, does that output the "real" (deduped) size increase for changed/new files or is it the nominal size?

**Comment by ThomasWaldmann** [comment]
Yes, everything **not** showing up in your dcsize listing are shared-by-2+ chunks.

borg diff: just have a look, i do not frequently use it myself.

dcsize PR should go against 1.2-maint branch. The stuff in master is different.
